### ISSUE TITLE
group_split() doesn't split groups on Ubuntu with dplyr 1.0.4

### ISSUE THREAD

**Issue by aftonsteps** [tidyverse/dplyr#5776]
`group_split()` seems to return an ungrouped full dataset with dplyr v1.0.4 on Ubuntu (Version 18.04.01 LTS). `group_split()` returns the correct set of group splits on Mac OS (Version 10.14.6). I have tibble v3.0.6 installed in both cases. I'm not sure if the OS is really the differentiator here but it seems to be. (EDIT: The same issue occurs on dplyr v1.0.3 for Ubuntu, and does not occur for Mac OS).
 
In this example, I'd expect `group_split()` to return a list of length 2, but on Ubuntu it returns a list of length 1 with no group splitting.

```r
library(magrittr)
test <- data.frame("a" = c(1,2), "b" = c(1,1))
test %>% dplyr::group_split(a)
```

Results on Ubuntu (list is too short and isn't grouped):

```
<list_of<
   tbl_df<
     a: double
     b: double
   >
 >[1]>
 [[1]]
 # A tibble: 2 x 2
       a     b
   <dbl> <dbl>
 1     1     1
 2     2     1
```

Results on Mac OS (list has 2 elements which is correct):

```
<list_of<
  tbl_df<
    a: double
    b: double
  >
>[2]>
[[1]]
# A tibble: 1 x 2
      a     b
  <dbl> <dbl>
1     1     1

[[2]]
# A tibble: 1 x 2
      a     b
  <dbl> <dbl>
1     2     1
```

**Comment by aftonsteps** [comment]
I think the problem might actually be with `ungroup()`. On the problem computer, if I call `ungroup()` on some object and assign the ungrouped data into a new variable, this _also_ ungroups and un-tibbles the original object. On the computer where `group_split()` works correctly, only the newly assigned variable points to an ungrouped object.

In `group_split()` specifically, at [this line](https://github.com/tidyverse/dplyr/blob/master/R/group_split.R#L113) in `group_split_impl()`, there's different behavior on each computer. On my Mac where things work as expected, the object `data` is a grouped tibble (class `grouped_df`, `tbl_df`, `tbl`, `data.frame`) both before and after the call to `ungroup()` at line 113. On the problem computer, it is a grouped tibble before, and a simple `data.frame` after.

I can also reproduce this behavior by using `ungroup()` directly:

```
test <- data.frame("a" = c(1,2), "b" = c(1,1)) %>% dplyr::group_by(a)
print(class(test))
shouldnt_matter <- dplyr::ungroup(test)
print(class(test))
```

On the Mac, where `group_split()` works correctly, this results in:

```
> test <- data.frame("a" = c(1,2), "b" = c(1,1)) %>% dplyr::group_by(a)

> print(class(test))
[1] "grouped_df" "tbl_df"     "tbl"        "data.frame"

> shouldnt_matter <- dplyr::ungroup(test)

> print(class(test))
[1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
```

On the Linux machine, this results in:

```
> test <- data.frame("a" = c(1,2), "b" = c(1,1)) %>% dplyr::group_by(a)

> print(class(test))
[1] "grouped_df" "tbl_df"     "tbl"        "data.frame"

> shouldnt_matter <- dplyr::ungroup(test)

> print(class(test))
[1] "data.frame"
```

The class of `test` changes to a simple `data.frame` in the problem case. I'm not really sure if the OS is truly the source of issue, but it's the only difference I can find between the two cases so far.

**Comment by romainfrancois** [comment]
Thanks for the details. It's very weird that `ungroup()` would alter its input.

**Comment by romainfrancois** [comment]
Does this also happen if you do these: 

```r
shouldnt_matter <- dplyr::as_tibble(test)
shouldnt_matter <- dplyr:::dplyr_vec_data(test)
shouldnt_matter <- as_tibble.grouped_df(test)
shouldnt_matter <- tibble:::as_tibble.data.frame(test)
```

I'm trying to find out how `test` is being modified in place when it should not

**Comment by aftonsteps** [comment]
The first three show the strange behavior where `test` gets untibbled, but the fourth leaves the original `test` as a `tibble`.

`test` is untibbled:
```r
> ## as_tibble
> test <- data.frame("a" = c(1,2), "b" = c(1,1)) %>% dplyr::group_by(a)
> print(class(test))
[1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
> shouldnt_matter <- dplyr::as_tibble(test)
> print(class(test))
[1] "data.frame"
> rm(test, shouldnt_matter)
```

```r
> ## dplyr_vec_data
> test <- data.frame("a" = c(1,2), "b" = c(1,1)) %>% dplyr::group_by(a)
> print(class(test))
[1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
> shouldnt_matter <- dplyr:::dplyr_vec_data(test)
> print(class(test))
[1] "data.frame"
> rm(test, shouldnt_matter)
```

```r
> ## as_tibble.grouped_df
> test <- data.frame("a" = c(1,2), "b" = c(1,1)) %>% dplyr::group_by(a)
> print(class(test))
[1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
> shouldnt_matter <- as_tibble.grouped_df(test)
> print(class(test))
[1] "data.frame"
> rm(test, shouldnt_matter)
```

`test` remains a `tibble`:
```r
> ## as_tibble.data.frame
> test <- data.frame("a" = c(1,2), "b" = c(1,1)) %>% dplyr::group_by(a)
> print(class(test))
[1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
> shouldnt_matter <- tibble:::as_tibble.data.frame(test)
> print(class(test))
[1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
> rm(test, shouldnt_matter)
```

**Comment by romainfrancois** [comment]
Thanks, so apparently the issue is somewhere in `dplyr_vec_data()`: 

```r
dplyr_vec_data <- function(x) {
  out <- vec_data(x)

  if (is.data.frame(x)) {
    new_data_frame(out, n = nrow(x))
  } else {
    out
  }
}
```

Either in `vec_data()` or `new_data_frame()`. Any idea if these would alter their input @lionel- @DavisVaughan  ?

@aftonsteps can you try with this please ? Which version of `vctrs` does this machine have ?

```r
shouldnt_matter <- vctrs::vec_data(test)
```

**Comment by DavisVaughan** [comment]
It would also be helpful to know what version of R you are using. I can't seem to reproduce on RStudio Cloud's Ubuntu 20.04.1 LTS instances. I tried with R 4.0.3 and the much older R 3.3.3

**Comment by aftonsteps** [comment]
Oops, yes, sorry for neglecting that. Here's the R version info:

```
platform       x86_64-pc-linux-gnu         
arch           x86_64                      
os             linux-gnu                   
system         x86_64, linux-gnu           
status                                     
major          3                           
minor          6.3                         
year           2020                        
month          02                          
day            29                          
svn rev        77875                       
language       R                           
version.string R version 3.6.3 (2020-02-29)
nickname       Holding the Windsock 
```

`vctrs` version is 0.3.6, which I think is the latest version. When I try `vec_data()`, I also get an untibbled `test`:

```r
> test <- data.frame("a" = c(1,2), "b" = c(1,1)) %>% dplyr::group_by(a)
> print(class(test))
[1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
> shouldnt_matter <- vctrs::vec_data(test)
> print(class(test))
[1] "data.frame"
```

**Comment by DavisVaughan** [comment]
Can you please run the following script for me, and post all of the results? You can copy the whole thing, then run `reprex::reprex()` in your R session while it is on your computer's clipboard and it should run it for you.

```r
library(vctrs)
library(dplyr)

make_test <- function() {
  tibble(x = 1, y = 2)
}
make_group_test <- function() {
  group_by(make_test(), x)
}

test <- make_test()
x <- vec_data(test)
test

test <- make_test()
x <- vec_proxy(test)
test

test <- make_test()
x <- new_data_frame(test, row.names = .row_names_info(test, 0L))
test

test <- make_group_test()
x <- vec_data(test)
test

test <- make_group_test()
x <- vec_proxy(test)
test

test <- make_group_test()
x <- new_data_frame(test, row.names = .row_names_info(test, 0L))
test
```

If you've never heard of a reprex before, you might want to start by reading the tidyverse.org [help page](https://www.tidyverse.org/help/).

You can install reprex by running (you may already have it, though, if you have the tidyverse package installed):
```r
install.packages("reprex")
```

**Comment by aftonsteps** [comment]
Yup! Thank you. 

Here's what happens. I get a warning when loading `vctrs` which is odd, but I could try with an earlier version of `vctrs` if you'd like.

```r
library(vctrs)
#> Warning: package 'vctrs' was built under R version 4.0.3
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following object is masked from 'package:vctrs':
#> 
#>     data_frame
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

make_test <- function() {
  tibble(x = 1, y = 2)
}
make_group_test <- function() {
  group_by(make_test(), x)
}

test <- make_test()
x <- vec_data(test)
test
#>   x y
#> 1 1 2

test <- make_test()
x <- vec_proxy(test)
test
#> # A tibble: 1 x 2
#>       x     y
#>   <dbl> <dbl>
#> 1     1     2

test <- make_test()
x <- new_data_frame(test, row.names = .row_names_info(test, 0L))
test
#>   x y
#> 1 1 2

test <- make_group_test()
x <- vec_data(test)
test
#>   x y
#> 1 1 2

test <- make_group_test()
x <- vec_proxy(test)
test
#> # A tibble: 1 x 2
#> # Groups:   x [1]
#>       x     y
#>   <dbl> <dbl>
#> 1     1     2

test <- make_group_test()
x <- new_data_frame(test, row.names = .row_names_info(test, 0L))
test
#>   x y
#> 1 1 2
```

<sup>Created on 2021-03-05 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0.9002)</sup>

**Comment by DavisVaughan** [comment]
Okay, I can at least reproduce this now. Knowing that you are on R 3.6.3 and seeing this was the key:

```r
library(vctrs)
#> Warning: package 'vctrs' was built under R version 4.0.3
```

How did you install vctrs? Possibly with RStudio Package Manager? Or directly from CRAN? It seems that you somehow have a version of vctrs that was built on a version of R that is much newer that your current version of R, and that is what is causing the issue. It is generally a little suspicious if you see an R package that was built on a version of R that is a major release (3 to 4) different from yours. Can you try reinstalling from source? Possibly using `remotes::install_cran("vctrs", force = TRUE)`.

For others, to reproduce this, go to rstudio.cloud, and:

- Open an R 4.0.3 instance, install vctrs
- Swap to R 3.6.3, go to `/home/rstudio-user/R`, find the 4.0 folder and copy the 4.0 version of vctrs over to the 3.6 folder.
- Then restart R 3.6.3 and try and run the reprex from above

**Comment by DavisVaughan** [comment]
Technical explanation of what is happening here:

vctrs was built with R 4.0.3, with reference counting. That means that `MAYBE_REFERENCED()`, defined as `!NO_REFERENCED()`, uses `REFCNT(x) == 0` to determine if a copy is required. https://github.com/wch/r-source/blob/6cd20a2b588200e0d934b06b07f14c34a3fc213e/src/include/Rinternals.h#L670-L685

vctrs was then loaded in R 3.6.3. This does not support reference counting, however, R core was testing it and had finalized most of the design. At that point, `REFCNT()` did exist, and was defined as `0` when the experimental reference counting support was not turned on https://github.com/wch/r-source/blob/6cd20a2b588200e0d934b06b07f14c34a3fc213e/src/include/Rinternals.h#L329-L335

The `REFCNT()` shown above is a macro, and in theory would have been inlined into the R 4.0.3 build of vctrs correctly when it was built with reference counting. However, this macro is only directly available when `USE_R_INTERNALS` is defined. Package authors generally get access to it through a function alias, which can be seen here https://github.com/wch/r-source/blob/92e2a4671e027ce7d048fd20c3c187657507b3e8/src/main/memory.c#L3763

Because it is a function in R's C API, and not a macro, when R 3.6.3 is used with this R 4.0.3 version of vctrs, calls to `REFCNT()` do indeed return `0`, improperly indicating that no copy is required, and resulting in the buggy behavior of `new_data_frame()`, which calls `r_clone_referenced()` (which uses `MAYBE_REFERENCED()`).

**Comment by aftonsteps** [comment]
Thank you for the technical explanation (and really sorry for bugging you, since this turns out to not be an issue with dplyr!).

I tried using `remotes::install_cran("vctrs", force = TRUE)`, and now the version is built under R3.6.3. 🎉 I also took a look at all the packages installed on this machine, and of 229 total packages, 111 of them were built under R 4.0.3. There is something odd here, because I'm not sure how that happened. To my knowledge, we only used `install.packages()` on this machine, but I'm just going to reinstall them all.

With the correct version of `vctrs`, the `dplyr` functions seem to no longer modify the argument in place, so now this returns the expected result:

```r
> test <- data.frame("a" = c(1,2), "b" = c(1,1)) %>% dplyr::group_by(a)
> print(class(test))
[1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
> shouldnt_matter <- dplyr::as_tibble(test)
> print(class(test))
[1] "grouped_df" "tbl_df"     "tbl"        "data.frame"
``` 

Thank you!
